src/jit/lowerxarch.cpp

```diff
@@ -1669,18 +1716,29 @@ void Lowering::ContainCheckMul(GenTreeOp* node)
             // Has a contained immediate operand.
             // Only 'other' operand can be marked as reg optional.
             assert(other != nullptr);
-            other->SetRegOptional();
+
+            bool isSafeToContainOther = (other == op1) ? isSafeToContainOp1 : isSafeToContainOp2;
+
+            if (isSafeToContainOther && IsSafeToContainMem(node, other))
+            {
+                other->SetRegOptional();
```

Comment:
This code is a little strange. See the following example which still wrongly marks the first operand as reg-optional, even though there is a type size mismatch:
```csharp
// Debug: Outputs 0
// Release: Outputs 6528000
struct S0
{
    public sbyte F1;
    public sbyte F2;
}

public class Program
{
    public static void Main()
    {
        S0 vr2 = new S0();
        vr2.F2--;
        int vr3 = System.Threading.Volatile.Read(ref vr2.F1) * 100;
        System.Console.WriteLine(vr3);
    }
}
```

the generated code is:
```asm
G_M5093_IG01:
       4883EC28             sub      rsp, 40
       33C0                 xor      rax, rax
       4889442420           mov      qword ptr [rsp+20H], rax

G_M5093_IG02:
       C644242000           mov      byte  ptr [rsp+20H], 0
       C644242100           mov      byte  ptr [rsp+21H], 0
       488D4C2421           lea      rcx, bword ptr [rsp+21H]
       FE09                 dec      byte  ptr [rcx]
      ; memory operand on this imul is marked reg-optional, even though the node is a byte node
       6B4C242064           imul     ecx, dword ptr [rsp+20H], 100
       E822FCFFFF           call     System.Console:WriteLine(int)
```
A few lines above there is a check for whether the type of the operand matches the node, and `memOp` is set to null because it doesn't. But then we get here and since there is a contained immediate, we redo the containment check without the type check and set it to reg-optional.

I realize this is probably not exactly #12398, but since you are making changes here anyway it might be nice to fix it too.

Reply:
@jakobbotsch I pushed a change that fixes this case. @CarolEidt PTAL